app/models/journaled/bulk_writer.rb

```diff
@@ -0,0 +1,44 @@
+class Journaled::BulkWriter
+  EVENTS_PER_DELIVERY = 500
+
+  def initialize(journaled_events:, app_name:, enqueue_opts: {})
+    @journaled_events = journaled_events.lazy
+    @app_name = app_name
+    @enqueue_opts = enqueue_opts
+  end
+
+  def journal!
+    serializers.each(&:serialize!)
+
+    enqueue_deliveries!
+  end
+
+  private
+
+  attr_reader :journaled_events, :app_name, :enqueue_opts
+
+  def enqueue_deliveries!
+    serializers.each_slice(EVENTS_PER_DELIVERY).each do |serializers|
+      enqueue_delivery!(serializers)
+    end
+  end
+
+  def enqueue_delivery!(serializers)
+    serialized_events = serializers.map(&:serialized_event)
+    partition_keys = serializers.map(&:journaled_partition_key)
+
+    Journaled.enqueue!(
+      Journaled::BulkDelivery.new(
+        records: serialized_events.zip(partition_keys),
+        app_name: app_name,
+      ),
+      enqueue_opts,
+    )
```

Comment:
Assuming that the serializers passed in here are in hash format so we are zipping here to get them to the array of arrays format? This kind of makes me think if zipping would also make sense in the bulk delivery above after all. 🤔 
Anyway, whichever way is the cleanest! 😅

Reply:
> Assuming that the serializers passed in here are in hash format

They are instances of `Journaled::Serializer` but ya they have the methods for both `serialized_event` and `journaled_partition_key` defined

Here is the commit where I switched it over if you want to see my other version!

https://github.com/WellSheet/journaled/pull/2/commits/e2d64d5e7a003645481d35212e02c465f76f5d0f